app/fork_manager.go

```diff
@@ -0,0 +1,65 @@
+package app
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type HardForkHandler interface {
+	GetTargetChainID() string
+	GetTargetHeight() int64
+	ExecuteHandler(ctx sdk.Context) error
+}
+
+type HardForkManager struct {
+	chainID    string
+	handlerMap map[int64][]HardForkHandler
+}
+
+// Create a new hard fork manager for a given chain ID.
+// This will filter out any handlers for other chain IDs,
+// and create a map that maps between target height and handlers to run for the given heights.
+func NewHardForkManager(chainID string) *HardForkManager {
+	return &HardForkManager{
+		chainID:    chainID,
+		handlerMap: make(map[int64][]HardForkHandler),
+	}
+}
+
+// This tries to register a handler with the hard fork manager.
+// If the handler's target chain ID doesn't match the chain ID of the manager,
+// this is a no-op and the handler is ignored.
+func (hfm *HardForkManager) RegisterHandler(handler HardForkHandler) {
+	if handler.GetTargetChainID() != hfm.chainID {
+		return
+	}
+	targetHeight := handler.GetTargetHeight()
+	if handlers, ok := hfm.handlerMap[targetHeight]; ok {
+		// TODO: refactor to use goutils InPlaceAppend once it's merged
+		handlers = append(handlers, handler)
+		hfm.handlerMap[targetHeight] = handlers
+	} else {
+		hfm.handlerMap[targetHeight] = []HardForkHandler{handler}
```

Comment:
is there a way to verify if targetHeight >= currHeight?

Reply:
I'm not certain, because this is performed at the chain startup in app.go so there isn't necessarily a `ctx` with a chain height present. is it an issue if old height handlers are stored in this map? we would expect the scale of this map to remain small, so IMO having historical ones shouldnt have negative effects

Reply:
Nope, its just for more protection in case we accidentally put in a smaller height. No worries if not possible